tmk_core/common/action_layer.c

```diff
@@ -110,94 +110,79 @@ void layer_debug(void)
 }
 #endif
 
-#if !defined(NO_ACTION_LAYER) && defined(PREVENT_STUCK_MODIFIERS)
-uint8_t source_layers_cache[(MATRIX_ROWS * MATRIX_COLS + 7) / 8][MAX_LAYER_BITS] = {0};
-
-void update_source_layers_cache(keypos_t key, uint8_t layer)
-{
-    const uint8_t key_number = key.col + (key.row * MATRIX_COLS);
-    const uint8_t storage_row = key_number / 8;
-    const uint8_t storage_bit = key_number % 8;
-
-    for (uint8_t bit_number = 0; bit_number < MAX_LAYER_BITS; bit_number++) {
-        source_layers_cache[storage_row][bit_number] ^=
-            (-((layer & (1U << bit_number)) != 0)
-             ^ source_layers_cache[storage_row][bit_number])
-            & (1U << storage_bit);
+#if !defined(NO_ACTION_LAYER) && defined(PREVENT_STUCK_KEYS)
+static uint8_t layers_cache[(MATRIX_ROWS * MATRIX_COLS + 7) / 8][MAX_LAYER_BITS];
+
+static void update_layers_cache(keypos_t key, uint8_t layer);
+static uint8_t read_layers_cache(keypos_t key);
+
+static void update_layers_cache(keypos_t key, uint8_t layer) {
+    uint8_t number = key.col + key.row * MATRIX_COLS;
+    uint8_t row = number / 8;
+    uint8_t mask = 1U << number % 8;
+    for (int8_t column = MAX_LAYER_BITS - 1; column >= 0; --column) {
+        mask_byte(&layers_cache[row][column], layer & _BV(column), mask);
```

Comment:
The code compiles for me. It seems that it's included in `pgmspace.h`, which is itself included in `action_macro.h`, which is itself included in `action_layer.c`.